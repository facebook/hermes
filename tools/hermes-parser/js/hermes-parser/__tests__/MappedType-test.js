/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

import {
  printForSnapshotESTree,
  parseForSnapshotESTree,
  printForSnapshotBabel,
  parseForSnapshotBabel,
} from '../__test_utils__/parse';

describe('MappedType', () => {
  describe('Basic', () => {
    const code = `
      type Mapped = {[key in keyof O]: O[key]};
    `;

    test('ESTree', async () => {
      expect(await parseForSnapshotESTree(code)).toMatchSnapshot();
      expect(await printForSnapshotESTree(code)).toBe(code.trim());
    });

    test('Babel', async () => {
      expect(await parseForSnapshotBabel(code)).toMatchSnapshot();
      expect(await printForSnapshotBabel(code)).toMatchInlineSnapshot(
        `"type Mapped = any;"`,
      );
    });
  });

  describe('Union', () => {
    const code = `
      type Mapped = {[key in keyof (O | Z)]: O[key]};
    `;

    test('ESTree', async () => {
      expect(await parseForSnapshotESTree(code)).toMatchSnapshot();
      expect(await printForSnapshotESTree(code)).toBe(code.trim());
    });

    test('Babel', async () => {
      expect(await parseForSnapshotBabel(code)).toMatchSnapshot();
      expect(await printForSnapshotBabel(code)).toMatchInlineSnapshot(
        `"type Mapped = any;"`,
      );
    });
  });
});
