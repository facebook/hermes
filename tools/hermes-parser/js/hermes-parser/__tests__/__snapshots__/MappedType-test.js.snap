// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MappedType Basic Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Mapped",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`MappedType Basic ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Mapped",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "callProperties": [],
        "exact": false,
        "indexers": [],
        "inexact": false,
        "internalSlots": [],
        "properties": [
          {
            "keyTparam": {
              "bound": null,
              "default": null,
              "name": "key",
              "type": "TypeParameter",
              "usesExtendsBound": false,
              "variance": null,
            },
            "optional": null,
            "propType": {
              "indexType": {
                "id": {
                  "name": "key",
                  "optional": false,
                  "type": "Identifier",
                  "typeAnnotation": null,
                },
                "type": "GenericTypeAnnotation",
                "typeParameters": null,
              },
              "objectType": {
                "id": {
                  "name": "O",
                  "optional": false,
                  "type": "Identifier",
                  "typeAnnotation": null,
                },
                "type": "GenericTypeAnnotation",
                "typeParameters": null,
              },
              "type": "IndexedAccessType",
            },
            "sourceType": {
              "argument": {
                "id": {
                  "name": "O",
                  "optional": false,
                  "type": "Identifier",
                  "typeAnnotation": null,
                },
                "type": "GenericTypeAnnotation",
                "typeParameters": null,
              },
              "type": "KeyofTypeAnnotation",
            },
            "type": "ObjectTypeMappedTypeProperty",
            "variance": null,
          },
        ],
        "type": "ObjectTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`MappedType Union Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Mapped",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`MappedType Union ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Mapped",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "callProperties": [],
        "exact": false,
        "indexers": [],
        "inexact": false,
        "internalSlots": [],
        "properties": [
          {
            "keyTparam": {
              "bound": null,
              "default": null,
              "name": "key",
              "type": "TypeParameter",
              "usesExtendsBound": false,
              "variance": null,
            },
            "optional": null,
            "propType": {
              "indexType": {
                "id": {
                  "name": "key",
                  "optional": false,
                  "type": "Identifier",
                  "typeAnnotation": null,
                },
                "type": "GenericTypeAnnotation",
                "typeParameters": null,
              },
              "objectType": {
                "id": {
                  "name": "O",
                  "optional": false,
                  "type": "Identifier",
                  "typeAnnotation": null,
                },
                "type": "GenericTypeAnnotation",
                "typeParameters": null,
              },
              "type": "IndexedAccessType",
            },
            "sourceType": {
              "argument": {
                "type": "UnionTypeAnnotation",
                "types": [
                  {
                    "id": {
                      "name": "O",
                      "optional": false,
                      "type": "Identifier",
                      "typeAnnotation": null,
                    },
                    "type": "GenericTypeAnnotation",
                    "typeParameters": null,
                  },
                  {
                    "id": {
                      "name": "Z",
                      "optional": false,
                      "type": "Identifier",
                      "typeAnnotation": null,
                    },
                    "type": "GenericTypeAnnotation",
                    "typeParameters": null,
                  },
                ],
              },
              "type": "KeyofTypeAnnotation",
            },
            "type": "ObjectTypeMappedTypeProperty",
            "variance": null,
          },
        ],
        "type": "ObjectTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;
